pkg/gcp/gcp.go

```diff
@@ -0,0 +1,178 @@
+	}
+
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner public ports are closed on GCP")
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(internalIngressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner internal ports are closed on GCP")
+
+	return nil
+}
+
+func (gc *gcpCloud) newFirewallRules(name string, ports []api.PortSpec, isInternal bool) (ingress, egress *compute.Firewall) {
+	ingressName, egressName := generateRuleNames(gc.infraID, name)
+
+	allowedPorts := []*compute.FirewallAllowed{}
+	for _, port := range ports {
+		allowedPorts = append(allowedPorts, &compute.FirewallAllowed{
+			IPProtocol: port.Protocol,
+			Ports:      []string{strconv.Itoa(int(port.Port))},
+		})
+	}
+
+	ingress = &compute.Firewall{
+		Name:      ingressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "INGRESS",
+		Allowed:   allowedPorts,
+	}
+
+	// if ports are internal, we make the ports are accessed in internal cluster and egress is not required
+	if isInternal {
+		ingress.TargetTags = []string{
+			fmt.Sprintf("%s-worker", gc.infraID),
+			fmt.Sprintf("%s-master", gc.infraID),
+		}
+		ingress.SourceTags = []string{
+			fmt.Sprintf("%s-worker", gc.infraID),
+			fmt.Sprintf("%s-master", gc.infraID),
+		}
+
+		return ingress, nil
+	}
+
+	return ingress, &compute.Firewall{
+		Name:      egressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "EGRESS",
+		Allowed:   allowedPorts,
+	}
+}
+
+// open expected ports by creating related firewall rule
+// - if the firewall rule is not found, we will create it
+// - if the firewall rule is found and changed, we will update it
+func (gc *gcpCloud) openPorts(rules ...*compute.Firewall) error {
+	for _, rule := range rules {
+		current, err := gc.client.GetFirewallRule(rule.Name)
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {
+			if err := gc.client.InsertFirewallRule(rule); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if !ruleChanged(current, rule) {
+			continue
+		}
+
+		if err := gc.client.UpdateFirewallRule(rule.Name, rule); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func generateRuleNames(infraID, name string) (ingressName, egressName string) {
+	return fmt.Sprintf("%s-%s-ingress", infraID, name), fmt.Sprintf("%s-%s-egress", infraID, name)
+}
+
+func ruleChanged(oldRule, newRule *compute.Firewall) bool {
+	if len(oldRule.Allowed) != len(newRule.Allowed) {
+		return true
+	}
+
+	if oldRule.Allowed[0].IPProtocol != newRule.Allowed[0].IPProtocol {
+		return true
+	}
+
+	return !reflect.DeepEqual(oldRule.Allowed[0].Ports, newRule.Allowed[0].Ports)
```

Comment:
This only checks the first allowed port, but what if there are several?

Also this function might be unnecessary based on my previous comment for `openPorts` (L152)